### PR TITLE
tests: Add "make check-valgrind" to run the unit tests under Valgrind

### PR DESCRIPTION
**Pull request by practicalswift**
Add `make check-valgrind` to run the unit tests under Valgrind.

~~Fix uninitialized read in `CWallet::CreateTransaction(...)` which is required for `make valgrind-check` to pass.~~ Update: Fixed by the merge of #17568.

Reviewers of this PR might be interested in the related PR #17633 ("tests: Add option --valgrind to run the functional tests under Valgrind").

Hopefully this will help kill the uninitialized read bug class :)

### CODE REVIEW

**Comment by fanquake** [comment]
Instead of https://github.com/bitcoin/bitcoin/pull/17639/commits/8c6edc2c7d069a3633042c842d74cd6a999c865f, you can cherry-pick from #17568, which already fixes the uninitialized read issue and adds a test.

**Comment by practicalswift** [comment]
@fanquake Done! :)

File: src/Makefile.test.include [hunk 1]
```diff
@@ -416,6 +416,31 @@ if EMBEDDED_UNIVALUE
...
+check-valgrind: $(TEST_BINARY) $(BENCH_BINARY)
+endif
+else
+if ENABLE_QT_TESTS
+check-valgrind: $(TEST_BINARY) test_bitcoin_qt
+else
+check-valgrind: $(TEST_BINARY)
+endif
+endif
+	@echo "Using the valgrind memory error detector: expect an ~50x slowdown, valgrind 3.14 or later required"
```
Comment (fanquake) [inline src/Makefile.test.include]: What creates the requirement for `3.14+`?

**Review by fanquake — COMMENTED** [review]
Concept ACK - Unfortunately Valgrind support on macOS is somewhat limited (even building the master branch).

File: src/Makefile.test.include (hunk 1 above)
Comment (practicalswift) [reply src/Makefile.test.include]: `--exit-on-first-error` was introduced in Valgrind 3.14.

File: src/Makefile.test.include [hunk 2]
```diff
@@ -416,6 +416,31 @@ if EMBEDDED_UNIVALUE
...
+	@echo "Running $(TEST_BINARY) under valgrind -- this will take a quite a while ..."
+	valgrind --suppressions=$(top_builddir)/contrib/valgrind.supp --gen-suppressions=all --exit-on-first-error=yes --error-exitcode=1 --quiet $(TEST_BINARY)
```
Comment (MarcoFalke) [inline src/Makefile.test.include]: I'd rather keep running the tests in parallel, even more so that valgrind causes a slowdown. See https://github.com/bitcoin/bitcoin/pull/17490/files#diff-3b1d470e34d4107835b2c4827cde6dc6R40

Comment (practicalswift) [reply src/Makefile.test.include]: I'm not sure how to do it in parallel in a way that a.) prints generates suppressions to the user, and b.) does not potentially introduce false positives due to resource exhaustion. Perhaps introducing parallelism can be tackled in a follow up PR?

**Comment by fjahr** [comment]
Concept ACK

Also unfortunately not able to test because of valgrind/macOS shenanigans at the moment.

**Comment by jonatack** [comment]
Would someone please add a `Review Club` label to this PR? Thanks :)

**Comment by practicalswift** [comment]
Welcome all review club members! Unfortunately I cannot attend your meeting today but you should know that I'm a huge fan of what you are doing: by reviewing you're helping to make Bitcoin Core harder, better, faster and stronger! :)

Looking forward to reading your feedback!

Remember: given enough eyeballs all bugs are shallow :)

File: src/Makefile.test.include [hunk 3]
```diff
@@ -416,6 +416,31 @@ if EMBEDDED_UNIVALUE
...
+if ENABLE_BENCH
+	@echo "Running $(BENCH_BINARY) -evals=1 -scaling=0 under valgrind -- this will take quite a while ..."
```
Comment (jonatack) [inline src/Makefile.test.include]: nits noticed while testing on the command line:

- I think "quite a" is a bit verbose here and in line 440 -- "this may take a while" on its own is fine

- s/while .../while.../

File: src/Makefile.test.include (hunk 1 above)
Comment (jonatack) [inline src/Makefile.test.include]: - suggestion: s/Using the valgrind memory error detector/Running tests with valgrind memcheck/

- "~50x slowdown": looking through the manual and SE comments, I see a wide range of perf reports. The manual's [Running your program under Memcheck](http://valgrind.org/docs/manual/quick-start.html#quick-start.mcrun) section states "Your program will run much slower (eg. 20 to 30 times) than normal, and use a lot more memory." Replies to [this SE question](https://stackoverflow.com/questions/29365611/does-running-valgrind-slow-down-my-application) both mention a factor of 10 slowdown.

    Perhaps "expect a 10-50x slowdown and ~2x memory usage, vagrind 3.14 or later required." 

    This also avoids the awkward "a" or "an" question before the "~50x".

In summary, perhaps something like:

"Running tests with valgrind memcheck: expect a 10-50x slowdown and 2x memory usage, vagrind 3.14 or later required"

**Review by jonatack — COMMENTED** [review]
Concept ACK 947a91870d568e1cb450d18739b6fa1cbb28e78b. Running `make check-valgrind` (with valgrind 3.14 on debian) outputs the following. After 15 minutes I see nothing further, so if the tests are running as expected, I'm not sure who will have the patience to run them. Perhaps I'm missing a configuration?
```
bitcoin ((HEAD detached at origin/pr/17639))$ make check-valgrind
make -C src/ check-valgrind
make[1]: Entering directory '/home/jon/projects/bitcoin/bitcoin/src'
make[2]: Entering directory '/home/jon/projects/bitcoin/bitcoin'
make[2]: Leaving directory '/home/jon/projects/bitcoin/bitcoin'
Using the valgrind memory error detector: expect an ~50x slowdown, valgrind 3.14 or later required
Running test/test_bitcoin under valgrind -- this will take a quite a while ...
valgrind --suppressions=../contrib/valgrind.supp --gen-suppressions=all --exit-on-first-error=yes --error-exitcode=1 --quiet test/test_bitcoin
Running 387 test cases...
```
EDIT: I debugged this extensively by running `valgrind --gen-suppressions=all --verbose --exit-on-first-error=yes --error-exitcode=1 --suppressions=contrib/valgrind.supp src/test/test_bitcoin --log_level=test_suite`. The issue for me was that the suppressions file needed additions to make it through the unit tests.

File: src/Makefile.test.include (hunk 2 above)
Comment (MarcoFalke) [reply src/Makefile.test.include]: a) The suppression should be printed when the test case fails, no?

b) I think we can't prevent OOM in general due to users setting too high `-j` in `make check`. Do you have any hints when this could lead to false positives instead of a killed process or frozen machine?

Comment (MarcoFalke) [reply src/Makefile.test.include]: Also, note that the same OOM "protection" does not exist for the functional tests when run with `--valgrind` or so

File: src/Makefile.test.include (hunk 3 above)
Comment (practicalswift) [reply src/Makefile.test.include]: Yes, that's better. Thanks! Now updated. Please re-review :)

File: src/Makefile.test.include (hunk 1 above)
Comment (practicalswift) [reply src/Makefile.test.include]: That's better. Thanks! Now updated. Please re-review :)

File: src/Makefile.test.include (hunk 2 above)
Comment (practicalswift) [reply src/Makefile.test.include]: I think these things would be nice to have but I'm afraid I won't have time to add that to this PR any time soon. I suggest we go with this simple version to start with and then someone can add parallellism in a follow-up PR. Makes sense? :)

**Comment by practicalswift** [comment]
@jonatack Thanks for reviewing. Would you mind re-reviewing the updated version? :)

File: src/Makefile.test.include [hunk 4]
```diff
@@ -416,6 +416,31 @@ if EMBEDDED_UNIVALUE
...
+else
+if ENABLE_QT_TESTS
+check-valgrind: $(TEST_BINARY) test_bitcoin_qt
+else
+check-valgrind: $(TEST_BINARY)
+endif
+endif
+	@echo "Using the valgrind memory error detector: expect a 10-50x slowdown and ~2x memory usage. valgrind 3.14 or later required."
+	@echo "Running $(TEST_BINARY) under valgrind -- this will take a a while..."
+	valgrind --suppressions=$(top_builddir)/contrib/valgrind.supp --gen-suppressions=all --exit-on-first-error=yes --error-exitcode=1 --quiet $(TEST_BINARY)
```
Comment (jonatack) [inline src/Makefile.test.include]: To have valgrind make it through the unit tests, I had to debug rather extensively my suppressions file by running  `valgrind --gen-suppressions=all --verbose --exit-on-first-error=yes --error-exitcode=1 --suppressions=contrib/valgrind.supp src/test/test_bitcoin --log_level=test_suite` and adding suppressions. It takes 60-90 minutes for a run, so it's a long process..

- Is the suppressions file for a project supposed to be portable across platforms, or is it normal for it to be a config adventure for each user?

- I found running valgrind directly more informative for debugging and feedback, particularly with flags like `--verbose` and `--log_level=test_suite`. I did not try running `make check-valgrind` with those options yet but wonder if it is a convenience feature, or adding other value that I'm misunderstanding? I do intend to retest and re-review.

Comment (practicalswift) [reply src/Makefile.test.include]: @jonatack 

I don't think you're doing anything wrong: Valgrind is super slow :)

> Is the suppressions file for a project supposed to be portable across platforms, or is it normal for it to be a config adventure for each user?

I think the current set of suppressions have been tested and updated by at least me and @MarcoFalke for various configurations, but more additions are welcome :)  Please submit your local changes upstreams :)

The suppressions are not necessarily portable across distributions, or across versions of the same distribution.

> I found running valgrind directly more informative for debugging and feedback, particularly with flags like --verbose and --log_level=test_suite. I did not try running make check-valgrind with those options yet but wonder if it is a convenience feature, or adding other value that I'm misunderstanding? I do intend to retest and re-review.

I usually run with `--log_level=test_suite` when debugging a found error, but I'm afraid enabling it by default for `make check-valgrind` would make the output too verbose. Generally I think `check`-type commands should be silent unless an error is found, and hopefully the most common scenarios is that of no Valgrind errors :)

**Comment by practicalswift** [comment]
Rebased! :)

**Comment by Binh0103** [comment]
..

**Review by Binh0103 — APPROVED** [review]

File: src/Makefile.test.include (hunk 4 above)
Comment (jonatack) [reply src/Makefile.test.include]: Retested this PR and wonder if it would not be better to remove the --quiet flag... `make check` isn't quiet either, `make check-valgrind` takes a long time (~90 minutes) to run so some output is reassuring, and the additional output is useful:
```
Using the valgrind memory error detector: expect a 10-50x slowdown and ~2x memory usage. valgrind 3.14 or later required.
Running test/test_bitcoin under valgrind -- this will take a a while...
valgrind --suppressions=../contrib/valgrind.supp --gen-suppressions=all --exit-on-first-error=yes --error-exitcode=1 test/test_bitcoin
==36831== Memcheck, a memory error detector
==36831== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==36831== Using Valgrind-3.16.0.GIT and LibVEX; rerun with -h for copyright info
==36831== Command: test/test_bitcoin
==36831== 
Running 391 test cases...
```

**Review by jonatack — COMMENTED** [review]
ACK 2a57a15 mod comment

File: src/Makefile.test.include [hunk 5]
```diff
@@ -613,6 +613,31 @@ if EMBEDDED_UNIVALUE
...
+if ENABLE_BENCH
```
Comment (jonatack) [inline src/Makefile.test.include]: Retested this PR and wonder if it would not be better to remove the --quiet flags... `make check` isn't quiet either, `make check-valgrind` takes a long time (~90 minutes) to run so some output is reassuring, and the additional output is useful:
```
Using the valgrind memory error detector: expect a 10-50x slowdown and ~2x memory usage. valgrind 3.14 or later required.
Running test/test_bitcoin under valgrind -- this will take a a while...
valgrind --suppressions=../contrib/valgrind.supp --gen-suppressions=all --exit-on-first-error=yes --error-exitcode=1 test/test_bitcoin
==36831== Memcheck, a memory error detector
==36831== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==36831== Using Valgrind-3.16.0.GIT and LibVEX; rerun with -h for copyright info
==36831== Command: test/test_bitcoin
==36831== 
Running 391 test cases...
```

Comment (practicalswift) [reply src/Makefile.test.include]: Adjusted accordingly :)

**Comment by practicalswift** [comment]
@jonatack No longer using `--quiet` as requested. Would you mind re-reviewing? :)

**Comment by jonatack** [comment]
Yes :) dumb question: how to have `make check-valgrind` use the ENABLE_BENCH and ENABLE_QT_TESTS conditional branches? Passing --enable-bench and --with-gui=yes doesn't seem to do it.

**Review by jonatack — COMMENTED** [review]
ACK 8942642e mod indentation nit (sorry for the iterative reviewing :)

I've never been able to get fully through any of the suites with valgrind yet because of missing suppressions (and may propose adding some). Therefore, to test this PR I needed to remove L629-631 (the unit tests) to see the bench tests, and then L632-635 (the bench tests) to see the qt tests. This is how I tested.

Edit: progress! valgrinded all the way through the bench and qt suites :)

File: src/Makefile.test.include [hunk 6]
```diff
@@ -613,6 +613,31 @@ if EMBEDDED_UNIVALUE
...
+else
+if ENABLE_QT_TESTS
+check-valgrind: $(TEST_BINARY) test_bitcoin_qt
+else
+check-valgrind: $(TEST_BINARY)
+endif
+endif
+	@echo "Using the valgrind memory error detector: expect a 10-50x slowdown and ~2x memory usage. valgrind 3.14 or later required."
+	@echo "Running $(TEST_BINARY) under valgrind -- this will take a a while..."
+	valgrind --suppressions=$(top_builddir)/contrib/valgrind.supp --gen-suppressions=all --exit-on-first-error=yes --error-exitcode=1 $(TEST_BINARY)
```
[the remark this answers is not in the document]
Comment (practicalswift) [reply src/Makefile.test.include]: This is the standard `Makefile` indentation we're following - see `%.cpp.test: %.cpp` indentation below as an example :)

**Comment by practicalswift** [comment]
Updated. Removed usage of `--exit-on-first-error` and thus the requirement of a very recent `valgrind` version.

After thinking about it I came to the conclusion that it is better to optimise for maximum number of `valgrind` testers than optimising for the rare case that a memory error is found.

Please re-review :)

**Comment by jonatack** [comment]
ACK 61fdd79 -- see https://gist.github.com/jonatack/1c0e0b9324ee4cc6c6bd6044c0b1c366 for output of running `make check-valgrind`.

Thanks to the updated valgrind suppressions in PR #18178 (review welcome), this PR now runs for me with no errors related to our dependencies.

**Comment by practicalswift** [comment]
Any chance we can move forward with this PR?

I'll leave it open for a while to allow for ACK:s and close if no interest is shown :)

**Comment by practicalswift** [comment]
Closing due to lack of interest :)
